File: frontend/src/features/delays/api/delays.ts
import { apiRequest } from "../../../lib/http/api-client";
import { booleanValue, jsonList, jsonObject, nonNegativeInteger, nullableText, text } from "../../../lib/http/json-contract";
import type { DelayReview, OperationalDelay, OperationalDelaysResult } from "../types/delays";

function review(value: unknown): DelayReview {
  const item = jsonObject(value, "review");
  return { id: text(item.id, "id"), decision: text(item.decision, "decision") as DelayReview["decision"], decided_at: text(item.decided_at, "decided_at"), decided_by: text(item.decided_by, "decided_by"), employee_occurrence_report_id: nullableText(item.employee_occurrence_report_id, "employee_occurrence_report_id"), note: nullableText(item.note, "note") };
}

function delay(value: unknown): OperationalDelay {
  const item = jsonObject(value, "delay");
  const source = jsonObject(item.source_fact, "source_fact");
  const schedule = item.schedule === null ? null : jsonObject(item.schedule, "schedule");
  return {
    delay_occurrence_id: text(item.delay_occurrence_id, "delay_occurrence_id"), collaborator_id: text(item.collaborator_id, "collaborator_id"), display_name: text(item.display_name, "display_name"), occurrence_date: text(item.occurrence_date, "occurrence_date"), occurrence_type: text(item.occurrence_type, "occurrence_type") as OperationalDelay["occurrence_type"], review_status: text(item.review_status, "review_status") as OperationalDelay["review_status"], counts_for_rv: booleanValue(item.counts_for_rv, "counts_for_rv"), observed_seconds: nonNegativeInteger(item.observed_seconds, "observed_seconds"), applied_limit_seconds: nonNegativeInteger(item.applied_limit_seconds, "applied_limit_seconds"),
    source_fact: { queue: text(source.queue, "queue"), started_at: text(source.started_at, "started_at"), ended_at: text(source.ended_at, "ended_at"), duration_seconds: nonNegativeInteger(source.duration_seconds, "duration_seconds"), pause_type: nullableText(source.pause_type, "pause_type") },
    schedule: schedule === null ? null : { planned_start: text(schedule.planned_start, "planned_start"), planned_end: nullableText(schedule.planned_end, "planned_end"), effective_origin: text(schedule.effective_origin, "effective_origin") },
    review: item.review === null ? null : review(item.review),
    possible_employee_occurrence_reports: jsonList(item.possible_employee_occurrence_reports, (entry) => { const report = jsonObject(entry, "report"); return { id: text(report.id, "id"), external_reference: text(report.external_reference, "external_reference"), external_collaborator_identity: text(report.external_collaborator_identity, "external_collaborator_identity"), submitted_at: text(report.submitted_at, "submitted_at"), occurrence_date: text(report.occurrence_date, "occurrence_date"), reason_text: text(report.reason_text, "reason_text") }; }, "possible_employee_occurrence_reports"),
  };
}

export function getOperationalDelays(filters: { competenceMonth: string; collaboratorId?: string; delayType?: string; reviewStatus?: string }, signal?: AbortSignal): Promise<OperationalDelaysResult> {
  const parameters = new URLSearchParams({ competence_month: filters.competenceMonth });
  if (filters.collaboratorId) parameters.set("collaborator_id", filters.collaboratorId);
  if (filters.delayType) parameters.set("delay_type", filters.delayType);
  if (filters.reviewStatus) parameters.set("review_status", filters.reviewStatus);
  return apiRequest(`/delays?${parameters}`, { signal, parse(value) { const root = jsonObject(value, "root"); return { competence_month: text(root.competence_month, "competence_month"), detected_count: nonNegativeInteger(root.detected_count, "detected_count"), pending_count: nonNegativeInteger(root.pending_count, "pending_count"), valid_count: nonNegativeInteger(root.valid_count, "valid_count"), corrected_count: nonNegativeInteger(root.corrected_count, "corrected_count"), items: jsonList(root.items, delay, "items") }; } });
}

export function createDelayReview(delayId: string, input: { decision: "valid" | "corrected"; employee_occurrence_report_id?: string; note?: string }): Promise<DelayReview> {
  return apiRequest(`/delays/${encodeURIComponent(delayId)}/reviews`, { method: "POST", body: JSON.stringify(input), parse: review });
}

File: backend/src/supervisor_ai/api/delays.py
from datetime import UTC, date, datetime, time
from typing import Annotated, Protocol
from uuid import uuid4

from fastapi import APIRouter, Path, Query
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from supervisor_ai.api.errors import error_response
from supervisor_ai.api.schemas import ErrorResponse
from supervisor_ai.application.errors import (
    DelayOccurrenceNotFound,
    DelayReviewConflict,
)
from supervisor_ai.application.persistence import DelayReview
from supervisor_ai.application.use_cases import (
    GetOperationalDelaysQuery,
    GetOperationalDelaysResult,
    RecordDelayReviewCommand,
)
from supervisor_ai.rules_engine.delays import DelayDecision

MVP_DELAY_REVIEW_ACTOR = "mvp-supervisor"


class OperationalDelayQueryContract(Protocol):
    def execute(
        self, query: GetOperationalDelaysQuery
    ) -> GetOperationalDelaysResult: ...


class DelayReviewContract(Protocol):
    def execute(self, command: RecordDelayReviewCommand) -> DelayReview: ...


class DelayReviewRequest(BaseModel):
    decision: DelayDecision
    employee_occurrence_report_id: str | None = Field(
        default=None, min_length=1, max_length=128
    )
    note: str | None = Field(default=None, min_length=1, max_length=2000)


def delays_router(
    query_service: OperationalDelayQueryContract, review_service: DelayReviewContract
) -> APIRouter:
    router = APIRouter(prefix="/delays", tags=["delays"])

    @router.get(
        "",
        response_model=None,
        responses={422: {"model": ErrorResponse}, 500: {"model": ErrorResponse}},
    )
    async def list_delays(
        competence_month: str,
        collaborator_id: Annotated[str | None, Query(min_length=1)] = None,
        delay_type: str | None = None,
        review_status: str | None = None,
    ) -> dict[str, object] | JSONResponse:
        try:
            month = date.fromisoformat(f"{competence_month}-01")
            result = query_service.execute(
                GetOperationalDelaysQuery(
                    month, collaborator_id, delay_type, review_status
                )
            )
        except ValueError:
            return error_response(
                422, "invalid_delay_filters", "Delay filters are invalid"
            )
        except Exception:
            return error_response(
                500, "internal_error", "Operational delays could not be retrieved"
            )
        return _result(result)

    @router.post(
        "/{delay_occurrence_id}/reviews",
        status_code=201,
        response_model=None,
        responses={
            404: {"model": ErrorResponse},
            409: {"model": ErrorResponse},
            422: {"model": ErrorResponse},
            500: {"model": ErrorResponse},
        },
    )
    async def create_review(
        delay_occurrence_id: Annotated[str, Path(min_length=1, max_length=128)],
        payload: DelayReviewRequest,
    ) -> dict[str, object] | JSONResponse:
        now = datetime.now(UTC)
        try:
            review = review_service.execute(
                RecordDelayReviewCommand(
                    f"delay-review-{uuid4()}",
                    delay_occurrence_id,
                    payload.decision,
                    now,
                    MVP_DELAY_REVIEW_ACTOR,
                    payload.employee_occurrence_report_id,
                    payload.note,
                )
            )
        except DelayOccurrenceNotFound:
            return error_response(
                404, "delay_occurrence_not_found", "Delay occurrence was not found"
            )
        except DelayReviewConflict:
            return error_response(
                409,
                "delay_review_conflict",
                "Review evidence does not match the delay occurrence",
            )
        except ValueError:
            return error_response(
                422, "invalid_delay_review", "Delay review is invalid"
            )
        except Exception:
            return error_response(
                500, "internal_error", "Delay review could not be recorded"
            )
        return _review(review)

    return router


def _result(result: GetOperationalDelaysResult) -> dict[str, object]:
    return {
        "competence_month": result.competence_month.strftime("%Y-%m"),
        "filters": {
            "collaborator_id": result.collaborator_id,
            "delay_type": result.delay_type,
            "review_status": result.review_status,
        },
        "detected_count": result.detected_count,
        "pending_count": result.pending_count,
        "valid_count": result.valid_count,
        "corrected_count": result.corrected_count,
        "items": [_item(item) for item in result.items],
    }


def _item(item: object) -> dict[str, object]:
    occurrence, source = item.occurrence, item.source_fact
    return {
        "delay_occurrence_id": occurrence.id,
        "collaborator_id": occurrence.collaborator_id,
        "display_name": item.display_name,
        "occurrence_date": occurrence.occurrence_date.isoformat(),
        "occurrence_type": occurrence.occurrence_type,
        "review_status": item.review_status,
        "counts_for_rv": item.counts_for_rv,
        "created_at": occurrence.created_at.isoformat(),
        "observed_seconds": occurrence.observed_seconds,
        "applied_limit_seconds": occurrence.applied_limit_seconds,
        "source_fact": {
            "type": source.source_fact_type,
            "id": source.source_fact_id,
            "source": source.source,
            "source_reference": source.source_reference,
            "source_extract_reference": source.source_extract_reference,
            "source_sheet": source.source_sheet,
            "source_row": source.source_row,
            "queue": source.queue,
            "started_at": source.started_at.isoformat(),
            "ended_at": source.ended_at.isoformat(),
            "duration_seconds": source.duration_seconds,
            "pause_type": source.pause_type,
        },
        "schedule": None if item.schedule is None else _schedule(item.schedule),
        "review": None if item.current_review is None else _review(item.current_review),
        "possible_employee_occurrence_reports": [
            {
                "id": report.id,
                "external_reference": report.external_reference,
                "source": report.source,
                "external_collaborator_identity": report.external_collaborator_identity,
                "submitted_at": report.submitted_at.isoformat(),
                "occurrence_date": report.occurrence_date.isoformat(),
                "reason_text": report.reason_text,
                "source_sheet": report.source_sheet,
                "source_row": report.source_row,
            }
            for report in item.possible_reports
        ],
    }


def _schedule(schedule: object) -> dict[str, object]:
    return {
        "planned_start": _time(schedule.planned_start),
        "planned_end": _time(schedule.planned_end),
        "effective_origin": schedule.effective_origin,
        "source_reference": schedule.source_reference,
        "source_sheet": schedule.source_sheet,
        "source_cell": schedule.source_cell,
    }


def _review(review: DelayReview) -> dict[str, object]:
    return {
        "id": review.id,
        "decision": review.decision,
        "decided_at": review.decided_at.isoformat(),
        "decided_by": review.decided_by,
        "employee_occurrence_report_id": review.employee_occurrence_report_id,
        "note": review.note,
        "created_at": review.created_at.isoformat(),
    }


def _time(value: time | None) -> str | None:
    return None if value is None else value.strftime("%H:%M:%S")
